apps/prairielearn/src/ee/pages/studentCourseInstanceUpgrade/studentCourseInstanceUpgrade.ts

```diff
@@ -17,7 +48,222 @@ router.get(
+        },
+        metadata: {
+          prairielearn_institution_id: institution.id,
+          prairielearn_institution_name: `${institution.long_name} (${institution.short_name})`,
+          prairielearn_course_id: course.id,
+          prairielearn_course_name: `${course.short_name}: ${course.title}`,
+          prairielearn_course_instance_id: course_instance.id,
+          prairielearn_course_instance_name: `${course_instance.long_name} (${course_instance.short_name})`,
+          prairielearn_user_id: user.user_id,
+        },
+        line_items: lineItems,
+        mode: 'payment',
+        success_url: `${urlBase}/success?session_id={CHECKOUT_SESSION_ID}`,
+        cancel_url: urlBase,
+      });
+
+      await insertStripeCheckoutSessionForUserInCourseInstance({
+        session_id: session.id,
+        institution_id: institution.id,
+        course_instance_id: course_instance.id,
+        user_id: user.user_id,
+        data: session,
+        plan_names: planNames,
+      });
+
+      if (!session.url) throw error.make(500, 'Stripe session URL not found');
+
+      res.redirect(session.url);
+    } else {
+      throw error.make(400, `Unknown action: ${req.body.__action}`);
+    }
+  }),
+);
+
+router.get(
+  '/success',
+  asyncHandler(async (req, res) => {
+    const institution = InstitutionSchema.parse(res.locals.institution);
+    const course = CourseSchema.parse(res.locals.course);
+    const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+
+    if (!req.query.session_id) throw error.make(400, 'Missing session_id');
+
+    const stripeSessionId = z.string().parse(req.query.session_id);
+
+    const localSession = await getStripeCheckoutSessionBySessionId(stripeSessionId);
+    if (!localSession) {
+      throw new Error(`Unknown Stripe session: ${stripeSessionId}`);
+    }
+
+    const stripe = getStripeClient();
+    const session = await stripe.checkout.sessions.retrieve(stripeSessionId);
+
+    // Verify that the session is associated with the current course instance
+    // and user. We shouldn't hit this during normal operations, but an attacker
+    // could try to replay a session ID from a different course instance or user.
+    if (
+      localSession.institution_id !== institution.id ||
+      localSession.course_instance_id !== course_instance.id ||
+      localSession.user_id !== res.locals.authn_user.user_id
+    ) {
+      throw error.make(400, 'Invalid session');
+    }
+
+    if (session.payment_status === 'paid') {
+      if (!localSession.plan_grants_created) {
+        // Create plan grants and mark the session as completed.
+        //
+        // Doing these mutations in a GET handler isn't great, but we have
+        // reasonable protection in place against replay attacks, and it would
+        // be difficult to perform a CSRF attack because the session must have
+        // been created in Stripe and must refer to the same user and course instance.
```

Comment:
👀

The alternative is to have "fulfillment" be purely asynchronous, which seems to be a horrible UX.

We could probably reorder things here so that we immediately short circuit if `localSession` indicates that plans were already granted and avoid hitting the Stripe API in that case, which could help prevent a malicious actor from forcing us to exceed out Stripe rate limits by repeatedly refreshing this page.

Reply:
I'm in favor of an immediate short-circuit at the start of processing.

Reply:
Would another possibility be for the `/success` handler to update the DB and then do an immediate 307 redirect to a `/show_result` page that only renders the result of the transaction? That might help with users reloading the page? It's not clear any of this is worth worrying about, of course, and a short-circuit at the top of the page code is functionally identical to this redirect in any case.

Reply:
Yeah, I'm not too worried about someone sitting there and refreshing the page. I'll switch the order so that we can short circuit without hitting Stripe's API.

Reply:
Switched the order and added explicit checking for completion in 6e13d86.